aiida/backends/managers/profile_access.py

```diff
@@ -0,0 +1,222 @@
+        and the function returns ``None``. The PID file contains the command of the process.
+
+        :raises ~aiida.common.exceptions.LockedProfileError: if the profile is locked.
+        """
+        error_message = (
+            f'process {self.process.pid} cannot access profile `{self.profile.name}`'
+            f'because it is being locked.'
+        )
+        self._raise_if_locked(error_message)
+
+        filepath_pid = self._dirpath_records / f'{self.process.pid}.pid'
+        filepath_tmp = self._dirpath_records / f'{self.process.pid}.tmp'
+
+        try:
+            # Write the content to a temporary file and then move it into place with an atomic operation.
+            # This prevents the situation where another process requests a lock while this file is being
+            # written: if that was to happen, when the locking process is checking for outdated records
+            # it will read the incomplete command, won't be able to correctly compare it with its running
+            # process, and will conclude the record is old and clean it up.
+            filepath_tmp.write_text(str(self.process.cmdline()))
+            os.rename(filepath_tmp, filepath_pid)
+
+            # Check again in case a lock was created in the time between the first check and creating the
+            # access record file.
+            error_message = (
+                f'profile `{self.profile.name}` was locked while process '
+                f'{self.process.pid} was requesting access.'
+            )
+            self._raise_if_locked(error_message)
+
+        except Exception as exc:
+            filepath_tmp.unlink(missing_ok=True)
+            filepath_pid.unlink(missing_ok=True)
+            raise exc
+
+    @contextlib.contextmanager
+    def lock(self):
+        """Request a lock on the profile for exclusive access.
+
+        This context manager should be used if exclusive access to the profile is required. Access will be granted if
+        the profile is currently not in use, nor locked by another process. During the context, the profile will be
+        locked, which will be lifted automatically as soon as the context exits.
+
+        :raises ~aiida.common.exceptions.LockingProfileError: if there are currently active processes using the profile.
+        :raises ~aiida.common.exceptions.LockedProfileError: if there currently already is a lock on the profile.
+        """
+        error_message = (
+            f'process {self.process.pid} cannot lock profile `{self.profile.name}` '
+            f'because it is already locked.'
+        )
+        self._raise_if_locked(error_message)
+
+        self._clear_stale_pid_files()
+
+        error_message = (
+            f'process {self.process.pid} cannot lock profile `{self.profile.name}` '
+            f'because it is being accessed.'
+        )
+        self._raise_if_active(error_message)
+
+        filepath = self._dirpath_records / f'{self.process.pid}.lock'
```

Comment:
Why does the _exclusive_ lock file to be PID specific?

Reply:
If we record the `PID` we can inform the user which process is locking when raising errors about not being able to access a profile and we can also perform some checks ourselves (i.e. decide that we will automatically check if the process is running and unlock if it is not). Putting it in the name was the easiest way.

Moreover, having separate locking files gives an easy way to prevent racing conditions: if by the end of the process of acquiring the lock there is more than one lock file, something "went wrong" (more than one process were simultaneously trying to acquire the lock) and we remove at least the one that correspond to the process checking. This makes it impossible for two processes to each think they acquired a lock for exclusive access.

Reply:
I don't follow this argument, isn't the whole point of using a single exclusive lock file to prevent race conditions? If it possible that two processes can obtain the same lock at the same time, then the file system does not support locking. If you need to have the PID, then just obtain the lock and then write the PID file into a distinct PID file.

To my understanding, the file locking procedure _should_ be an atomic operation, whereas checking for files, creating a file, and then checking for file existence again is not.

There is also https://pypi.org/project/pid/ which might support that exact use case.

Reply:
> Moreover, having separate locking files gives an easy way to prevent racing conditions: 

@csadorf is right here. The whole point of a lock file is to prevent race conditions, so the `filelock` library would naturally protect against this.

Reply:
I am a bit lost in this thread now. I don't know if maybe a "lock file" is a more technical term with very specific details of how it is implemented; I'm using files to track access to the AiiDA profile and provide a context in which I guarantee no other process was accessing it.

My design indeed can't get completely rid of the racing condition, but it exchanges the situation of two competing processes both thinking they acquired the lock for a situation in which they both think the other one beat them to it and except. I think this should be good enough for our use case since it effectively prevents any corruption of the data and the user can just try again (I can't imagine a case where there are multiple processes trying to lock the profile **by design**). Using `filelock` would solve this but I still think this does not outweigh the costs of including an extra dependency and extra complexity for coordinating it with our other `.pid` tracking files (and having to keep track separately of the id of the locking process).

> There is also https://pypi.org/project/pid/ which might support that exact use case.

Mmm maybe? The "stale detection" line sounds interesting, but honestly I can't even understand what exactly does this do from the documentation they have (which seems to be reduced to that small readme).

Reply:
I think the requesting access part would remain a bespoke solution, but at least the `lock` implementation could be potentially simplified when using the `filelock` library:
```python
@contextlib.contextmanager
def lock(self):
    self._clear_stale_pid_files()
    try:
        with FileLock(self._dirpath_records / f'{self.process.pid}.lock'):
            self._raise_if_active(f'cannot lock profile `{self.profile.name}` because it is being accessed.')
            yield
    except LockingProfileError as exception:
        raise
    except Timeout as exception:
        raise LockedProfileError(f'cannot lock profile `{self.profile.name}` because it is locked.') from exc
 ```
It is significantly simpler and probably more robust across platforms.

Reply:
Mmm correct me if I wrong but I think what you are proposing doesn't work because each process will try to acquire a different lock file with its own name. In order for this to work we need to have all processes use the same locking file, in which case we can no longer store the `pid` as part of the name and we need to track that separately (as you pointed out in the other thread of this review).

I remain unconvinced about the cost-benefit balance of using `filelock`, but if you guys still think that it makes things simpler (even considering the extra file for keeping the `pid` info) then I can implement the necessary changes. I would be more strongly against completely removing the feature of printing the `pid` just for simplicity's sake (not that anyone suggested this yet, but just preemptive clarification).

Reply:
> Mmm correct me if I wrong but I think what you are proposing doesn't work because each process will try to acquire a different lock file with its own name.

Of course you are right, just a sloppy mistake because I just sketched this quickly to get the gist. It is true that we would probably have to give up including the pid in the lock file. I think this is fine personally. If it is locked, it should be evident what process is locking since there can only be one, and here there should be automatic cleanup and so a lock file by a dead process seems unlikely.

Anyway, I was just trying to quickly actually work this out so we can see what would be feasible to change so we can put this to bed. Without trying I fear this PR will always stay in stalemate.

Reply:
> I think this is fine personally. If it is locked, it should be evident what process is locking since there can only be one, and here there should be automatic cleanup and so a lock file by a dead process seems unlikely.

Mmm I'm not sure I agree. What do you mean by "it should be evident (...) since there can only be one"? There can only be one process locking **each profile**, but if you have to work with multiple profiles at the same time I am not sure it is so easy to determine which verdi instance is working with each profile by just running top or something similar, is it?

Reply:
> Anyway, I was just trying to quickly actually work this out so we can see what would be feasible to change so we can put this to bed. Without trying I fear this PR will always stay in stalemate.

I agree with the sentiment, I think I am trying to thoroughly consider the options you are putting forward and not just finding fault gratuitously. Maybe I am missing some information that makes the sorting of these obstacles trivial to you, but I'm trying to point out some consequences that I think are key to the functionality.

Reply:
Just to be clear, I am also not trying to antagonize here. I am just trying to help by trying to see if the suggestion of @chrisjsewell and @csadorf could work. The best way to do this and to show that we are taking it seriously, is to actually try it and propose an alternate implementation. Making it concrete is the best way to visualize any problems and could help us finalize the decision with the design. Otherwise, if we stick with just discussions we never move forward. But I think we have done enough now and I think we should just stick with this design then. Unless the others really think that having the additional information of which PIDs are blocking is not worth the additional complexity of the design and forcing a custom implementation.

Reply:
> All of these use cases could certainly be implemented with aforementioned libraries.

This is what you keep saying, but then when we describe potential reasons why this is not possible, we don't get a specific response whether you agree with that analysis. I tried to fix this by making it more concrete, going out of my way to provide a potential implementation on a PR that is not mine, but still we get a repetition of platitudes that it should be possible.

> I am starting to get somewhat frustrated needing to defend that suggestion over and over again.

Well that makes two of us. I am not asking you to defend it, I am asking you to look at our analysis of trying to implement your suggestion and the limitations that we think are there, and either provide a clear solution to those limitations or accept that they are there and so sign off on the design. Again, I am doing this in order to not discard your suggestion without proper consideration, out of respect to you and the time you took to review the concept of this PR, not to cause you grief. Just reiterating that "you think it should be possible with the library" is not helping.

Reply:
> > All of these use cases could certainly be implemented with aforementioned libraries.
> 
> This is what you keep saying, but then when we describe potential reasons why this is not possible, we don't get a specific response whether you agree with that analysis. I tried to fix this by making it more concrete, going out of my way to provide a potential implementation on a PR that is not mine, but still we get a repetition of platitudes that it should be possible.

I do not understand what you are expecting from me here. Do I have to implement it myself? I think I provided a clear analysis on what I interpret we are trying to implement here and made suggestion on how to improve the language and abstraction as well as provided guidance as to how I would implement it. When I inquired about specific obstacles the presented arguments were not convincing.

> > I am starting to get somewhat frustrated needing to defend that suggestion over and over again.
> 
> Well that makes two of us. I am not asking you to defend it, I am asking you to look at our analysis of trying to implement your suggestion and the limitations that we think are there, and either provide a clear solution to those limitations or accept that they are there and so sign off on the design. Again, I am doing this in order to not discard your suggestion without proper consideration, out of respect to you and the time you took to review the concept of this PR, not to cause you grief. Just reiterating that "you think it should be possible with the library" is not helping.

Again, not quite sure what is expected here despite providing an actual implementation. You have already received my approval to go ahead multiple times.

I apologize for not providing a constructive review and will withdraw from the discussion. I am very sorry that I wasted everybody's time with my suggestion.

Reply:
> I apologize for not providing a constructive review and will withdraw from the discussion. I am very sorry that I wasted everybody's time with my suggestion.

I am sorry if anything I said gave the impression that this was my stance, it is not at all. I really appreciate you taking the time to give feedback on this, and I truly think it was very valuable even if in the end I might opt or push to not apply all of it. I assure I did not just dismissed it: I really tried looking into the `filelock` library and playing around with it, but couldn't figure out a non-forceful way to leverage it for our specific conditions. I wanted to respond to all your points not as an attack to them, nor expecting you to defend the suggestions, but because they were perfectly valid points and thus merited an explanation of why, for this particular use case, I might prefer to decide against them.

I also tried my best to start using the proposed terminology throughout the discussion, is just that it is a bit hard for me to immediately understand how exactly the terms are intended and what is the extent of their influence (e.g. is "exclusive access" a qualifier when talking about locking? does it completely replace "locking"? can I still talk about lock files or how do I refer to the files I use for tracking this now?). Again, this is not to say it was on you to clarify all this right away, I'm just trying to explain why I may not have seem to fully adopt these semantics right away.

Lastly I hope my comments on the `pid` library's documentation didn't contribute to this. It was meant to express my frustration at the library itself, not at your recommendation of it which was very pertinent. Apologies if it seemed that way.

I think @giovannipizzi may be right, this has already eroded all of us a bit and it might be better to just go ahead and merge it. If at some point someone wishes to revisit incorporating any of these libraries with some fresh eyes or improving the feature in any other way, I would be more than happy to help.